llvm/lib/MC/MCRegisterInfo.cpp

```diff
@@ -151,24 +151,28 @@ int MCRegisterInfo::getDwarfRegNum(MCRegister RegNum, bool isEH) const {
   const DwarfLLVMRegPair *I = std::lower_bound(M, M+Size, Key);
   if (I == M+Size || I->FromReg != RegNum)
     return -1;
-  return I->ToReg;
+  // Consumers need to be able to detect -1 and -2, but at various points
+  // the numbers move between unsigned and signed representations, as well as
+  // between 32- and 64-bit representations. We need to convert first to int
+  // before int64_t for proper sign handling.
+  return int64_t(int(I->ToReg));
```

Comment:
what about doing the following:
```
auto reg = I->ToReg;
switch (reg) {
  case -1: return -1;
  case -2: return -2;
  default: return reg;
}
```

Reply:
That might be good.  I'm not certain there are not other negative values currently being used as signal values, but if that switch passes the tests then it's probably fine.  I'm not certain that's necessarily better, though.

Reply:
Then, what about `int64_t(int32_t(I->ToReg))`?

Reply:
All of the existing code dealing with these register numbers uses the types `int` and `unsigned`.  The `int64_t(int(I->ToReg))` conversion seems more consistent with other usage than `int64_t(int32_t(I->ToReg))`.